File: backend/src/agent/voiceAgent.ts
import { EventEmitter } from 'events';
import { DeepgramService } from '../services/deepgram';
import { GeminiService } from '../services/gemini';
import { ElevenLabsService } from '../services/elevenlabs';
import { SessionState, ConversationMessage, STTResult } from '../types';
import { GISA_INITIAL_MESSAGE, GISA_SYSTEM_PROMPT } from './gisaPrompt';

export class VoiceAgent extends EventEmitter {
  private sttService: DeepgramService;
  private llmService: GeminiService;
  private ttsService: ElevenLabsService;
  private sessionState: SessionState;
  private interimTranscript: string = '';
  private isProcessing: boolean = false;

  constructor(sessionId: string) {
    super();

    this.sttService = new DeepgramService();
    this.llmService = new GeminiService();
    this.ttsService = new ElevenLabsService();

    this.sessionState = {
      sessionId,
      conversationHistory: [
        {
          role: 'system',
          content: GISA_SYSTEM_PROMPT,
          timestamp: Date.now(),
        },
      ],
      currentPhase: 'FASE_1',
      ucValidated: false,
      startTime: Date.now(),
    };

    this.setupSTTListeners();
  }

  private setupSTTListeners(): void {
    this.sttService.on('transcript', async (result: STTResult) => {
      if (result.isFinal) {
        console.log('📝 Final transcript:', result.transcript);
        this.interimTranscript = '';

        if (!this.isProcessing) {
          await this.processUserInput(result.transcript);
        }
      } else {
        this.interimTranscript = result.transcript;
        console.log('💭 Interim:', this.interimTranscript);
      }
    });

    this.sttService.on('error', (error) => {
      console.error('❌ STT Error:', error);
      this.emit('error', error);
    });
  }

  async initialize(): Promise<void> {
    try {
      console.log('🚀 Initializing voice agent...');

      // Start STT streaming
      await this.sttService.startStreaming();

      // Generate and send initial greeting
      await this.sendInitialGreeting();

      console.log('✅ Voice agent initialized');
    } catch (error) {
      console.error('❌ Failed to initialize voice agent:', error);
      throw error;
    }
  }

  private async sendInitialGreeting(): Promise<void> {
    try {
      // Add to conversation history
      this.sessionState.conversationHistory.push({
        role: 'assistant',
        content: GISA_INITIAL_MESSAGE,
        timestamp: Date.now(),
      });

      // Generate audio
      const audioBuffer = await this.ttsService.textToSpeech(
        GISA_INITIAL_MESSAGE
      );

      // Emit audio to be sent via LiveKit
      this.emit('audio', audioBuffer);
    } catch (error) {
      console.error('❌ Failed to send initial greeting:', error);
      throw error;
    }
  }

  private async processUserInput(transcript: string): Promise<void> {
    if (this.isProcessing || !transcript.trim()) {
      return;
    }

    this.isProcessing = true;

    try {
      console.log('🎯 Processing user input:', transcript);

      // Add user message to history
      this.sessionState.conversationHistory.push({
        role: 'user',
        content: transcript,
        timestamp: Date.now(),
      });

      // Get LLM response
      const llmResponse = await this.llmService.generateResponse(
        this.sessionState.conversationHistory
      );

      console.log('🤖 LLM Response:', llmResponse.text);

      // Update session state based on metadata
      if (llmResponse.metadata) {
        if (llmResponse.metadata.phase) {
          this.sessionState.currentPhase = llmResponse.metadata.phase;
        }

        // Check if UC was validated
        if (
          llmResponse.text.includes('validei') ||
          llmResponse.text.includes('Perfeito')
        ) {
          this.sessionState.ucValidated = true;
        }
      }

      // Add assistant response to history
      this.sessionState.conversationHistory.push({
        role: 'assistant',
        content: llmResponse.text,
        timestamp: Date.now(),
      });

      // Generate speech
      const audioBuffer = await this.ttsService.textToSpeech(
        llmResponse.text
      );

      // Emit audio to be sent via LiveKit
      this.emit('audio', audioBuffer);

      // Emit transcript for UI
      this.emit('response', {
        text: llmResponse.text,
        metadata: llmResponse.metadata,
      });

    } catch (error) {
      console.error('❌ Error processing user input:', error);
      this.emit('error', error);
    } finally {
      this.isProcessing = false;
    }
  }

  processAudio(audioData: Buffer): void {
    this.sttService.sendAudio(audioData);
  }

  getSessionState(): SessionState {
    return { ...this.sessionState };
  }

  async shutdown(): Promise<void> {
    console.log('🛑 Shutting down voice agent...');
    this.sttService.close();
    this.removeAllListeners();
  }
}

File: backend/src/agent/voice_agent.py
"""Voice Agent orchestrator."""
import asyncio
import time
from typing import Optional
from ..models import SessionState, ConversationMessage, STTResult
from ..services.deepgram import DeepgramService
from ..services.gemini import GeminiService
from ..services.elevenlabs import ElevenLabsService
from .gisa_prompt import GISA_INITIAL_MESSAGE, GISA_SYSTEM_PROMPT


class VoiceAgent:
    """Voice agent that orchestrates STT, LLM, and TTS."""

    def __init__(self, session_id: str):
        """Initialize voice agent."""
        self.session_id = session_id
        self.stt_service = DeepgramService()
        self.llm_service = GeminiService()
        self.tts_service = ElevenLabsService()

        self.session_state = SessionState(
            session_id=session_id,
            conversation_history=[
                ConversationMessage(
                    role='system',
                    content=GISA_SYSTEM_PROMPT,
                    timestamp=time.time(),
                )
            ],
            current_phase='FASE_1',
            uc_validated=False,
            start_time=time.time(),
        )

        self.interim_transcript = ''
        self.is_processing = False
        self.on_audio_callback: Optional[callable] = None
        self.on_response_callback: Optional[callable] = None

    async def initialize(self):
        """Initialize the voice agent."""
        try:
            print('🚀 Initializing voice agent...')

            # Start STT streaming
            await self.stt_service.start_streaming()

            # Set up STT callbacks
            self.stt_service.on_transcript = self._handle_transcript
            self.stt_service.on_error = self._handle_error

            # Generate and send initial greeting
            await self._send_initial_greeting()

            print('✅ Voice agent initialized')

        except Exception as e:
            print(f'❌ Failed to initialize voice agent: {e}')
            raise

    async def _send_initial_greeting(self):
        """Send initial greeting."""
        try:
            # Add to conversation history
            self.session_state.conversation_history.append(
                ConversationMessage(
                    role='assistant',
                    content=GISA_INITIAL_MESSAGE,
                    timestamp=time.time(),
                )
            )

            # Generate audio
            audio_bytes = await self.tts_service.text_to_speech(GISA_INITIAL_MESSAGE)

            # Emit audio
            if self.on_audio_callback:
                await self.on_audio_callback(audio_bytes)

        except Exception as e:
            print(f'❌ Failed to send initial greeting: {e}')
            raise

    async def _handle_transcript(self, result: dict):
        """Handle transcript from STT."""
        if result['is_final']:
            print(f"📝 Final transcript: {result['transcript']}")
            self.interim_transcript = ''

            if not self.is_processing:
                await self._process_user_input(result['transcript'])
        else:
            self.interim_transcript = result['transcript']
            print(f"💭 Interim: {self.interim_transcript}")

    async def _handle_error(self, error):
        """Handle STT error."""
        print(f'❌ STT Error: {error}')

    async def _process_user_input(self, transcript: str):
        """Process user input."""
        if self.is_processing or not transcript.strip():
            return

        self.is_processing = True

        try:
            print(f'🎯 Processing user input: {transcript}')

            # Add user message to history
            self.session_state.conversation_history.append(
                ConversationMessage(
                    role='user',
                    content=transcript,
                    timestamp=time.time(),
                )
            )

            # Get LLM response
            llm_response = await self.llm_service.generate_response(
                self.session_state.conversation_history
            )

            print(f'🤖 LLM Response: {llm_response.text}')

            # Update session state based on metadata
            if llm_response.metadata:
                if 'phase' in llm_response.metadata:
                    self.session_state.current_phase = llm_response.metadata['phase']

                # Check if UC was validated
                if (
                    'validei' in llm_response.text.lower()
                    or 'perfeito' in llm_response.text.lower()
                ):
                    self.session_state.uc_validated = True

            # Add assistant response to history
            self.session_state.conversation_history.append(
                ConversationMessage(
                    role='assistant',
                    content=llm_response.text,
                    timestamp=time.time(),
                )
            )

            # Generate speech
            audio_bytes = await self.tts_service.text_to_speech(llm_response.text)

            # Emit audio
            if self.on_audio_callback:
                await self.on_audio_callback(audio_bytes)

            # Emit response for UI
            if self.on_response_callback:
                await self.on_response_callback({
                    'text': llm_response.text,
                    'metadata': llm_response.metadata,
                })

        except Exception as e:
            print(f'❌ Error processing user input: {e}')
        finally:
            self.is_processing = False

    async def process_audio(self, audio_data: bytes):
        """Process incoming audio."""
        await self.stt_service.send_audio(audio_data)

    def get_session_state(self) -> SessionState:
        """Get current session state."""
        return self.session_state

    async def shutdown(self):
        """Shutdown the voice agent."""
        print('🛑 Shutting down voice agent...')
        await self.stt_service.close()
